onnxruntime/contrib_ops/cuda/bert/multihead_attention.cc

```diff
@@ -19,22 +21,26 @@ namespace onnxruntime {
 namespace contrib {
 namespace cuda {
 
-#define REGISTER_KERNEL_TYPED(T)                                  \
-  ONNX_OPERATOR_TYPED_KERNEL_EX(                                  \
-      MultiHeadAttention,                                         \
-      kMSDomain,                                                  \
-      1,                                                          \
-      T,                                                          \
-      kCudaExecutionProvider,                                     \
-      (*KernelDefBuilder::Create())                               \
-          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>()), \
-      MultiHeadAttention<T>);
-
-REGISTER_KERNEL_TYPED(float)
-REGISTER_KERNEL_TYPED(MLFloat16)
-
-template <typename T>
-MultiHeadAttention<T>::MultiHeadAttention(const OpKernelInfo& info)
+#define REGISTER_KERNEL_TYPED(T, QK)                               \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                   \
+      MultiHeadAttention,                                          \
+      kMSDomain,                                                   \
+      1,                                                           \
+      T##_##QK,                                                    \
+      kCudaExecutionProvider,                                      \
+      (*KernelDefBuilder::Create())                                \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>())   \
+          .TypeConstraint("QK", DataTypeImpl::GetTensorType<QK>()) \
+          .InputMemoryType(OrtMemTypeCPUInput, 8),                 \
+      MultiHeadAttention<T, QK>);
+
+REGISTER_KERNEL_TYPED(float, float)
+REGISTER_KERNEL_TYPED(float, MLFloat16)
+REGISTER_KERNEL_TYPED(MLFloat16, float)
+REGISTER_KERNEL_TYPED(MLFloat16, MLFloat16)
```

Comment:
Is there need to support T=MLFloat16 and QK=MLFloat16?
I think QK is float in most cases.